# PART 1

In [1]:
import numpy as np
import pandas as pd 
import folium 
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [2]:
table_url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df = pd.read_html(table_url)[0]
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [3]:
sum(df['Postal Code'] == 'Not assigned') # sum = 0 means there are not rows with that value in that column

0

In [4]:
Bor_nan = df['Borough'] == 'Not assigned'
Neigh_nan = df['Neighbourhood'] == 'Not assigned'

In [5]:
df = df[Bor_nan == False]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [6]:
# Checks:
sum(df['Borough'] == 'Not assigned')

0

In [7]:
df[df['Postal Code'] == 'M5A']

Postal Code           Borough              Neighbourhood
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Neighbourhoods are already grouped

In [8]:
sum(df['Neighbourhood'] == 'Not assigned')

0

In [9]:
df.shape

(103, 3)

# PART 2 

In [10]:
import geocoder 
## Test with geocoder --> Doesn´t work 
postal_code = 'M5G'
lat_lng_coord = None 
counter = 0 
while lat_lng_coord is None:
    g = geocoder.google(f'{postal_code}, Toronto, Ontario')
    lat_lng_coord = g.latlng
    counter += 1
    if counter >= 10:
        break
    
#latitude = lat_lng_coord[0]
#longitude = lat_lng_coord[1]
#print(latitude,longitude)

In [11]:
coordinates_df = pd.read_csv('Geospatial_Coordinates.csv')
coordinates_df.set_index('Postal Code', inplace=True)
coordinates_df.head()

Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476

In [12]:
coordinates_df.loc['M1B']

Latitude     43.806686
Longitude   -79.194353
Name: M1B, dtype: float64

In [13]:
df['Postal Code'].apply(lambda postal_code: coordinates_df.loc[postal_code][0])
df['latitude'] = df['Postal Code'].apply(lambda postal_code: coordinates_df.loc[postal_code][0])
df['longitude'] = df['Postal Code'].apply(lambda postal_code: coordinates_df.loc[postal_code][1])
df.head()

Postal Code           Borough                                Neighbourhood  \
2         M3A        North York                                    Parkwoods   
3         M4A        North York                             Victoria Village   
4         M5A  Downtown Toronto                    Regent Park, Harbourfront   
5         M6A        North York             Lawrence Manor, Lawrence Heights   
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    latitude  longitude  
2  43.753259 -79.329656  
3  43.725882 -79.315572  
4  43.654260 -79.360636  
5  43.718518 -79.464763  
6  43.662301 -79.389494

## Part 2: Explore and cluster the neighborhoods in Toronto

In [14]:
Toronto_df = df[df['Borough'].str.contains('Toronto')]
Toronto_df.reset_index(inplace=True)
Toronto_df.head()

index Postal Code           Borough  \
0      4         M5A  Downtown Toronto   
1      6         M7A  Downtown Toronto   
2     13         M5B  Downtown Toronto   
3     22         M5C  Downtown Toronto   
4     30         M4E      East Toronto   

                                 Neighbourhood   latitude  longitude  
0                    Regent Park, Harbourfront  43.654260 -79.360636  
1  Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                     Garden District, Ryerson  43.657162 -79.378937  
3                               St. James Town  43.651494 -79.375418  
4                                  The Beaches  43.676357 -79.293031

In [15]:
Toronto_df.shape

(39, 6)

There are 39 postal codes in Toronto found in the wikipedia data.

In [16]:
# Toronto Coordinates 
Tor_lat = 43.6529
Tor_lon = -79.3849

In [17]:
map_Toronto = folium.Map(location=[Tor_lat, Tor_lon], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['latitude'], df['longitude'], df['Borough'], df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

In [18]:
CLIENT_ID = 'CP0ATHJXNULXOKBLBZFRPUZUCZNMVJ51AJVVYWZB0F3GNDDS' # your Foursquare ID
CLIENT_SECRET = 'Q421JSGYLSTZZOTNM1KN3HO5KWWZDCE1U5J1PKUE2LUZUDC4' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CP0ATHJXNULXOKBLBZFRPUZUCZNMVJ51AJVVYWZB0F3GNDDS
CLIENT_SECRET:Q421JSGYLSTZZOTNM1KN3HO5KWWZDCE1U5J1PKUE2LUZUDC4


In [19]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [20]:
Toronto_venues = getNearbyVenues(names=df['Neighbourhood'],
                                   latitudes=df['latitude'],
                                   longitudes=df['longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [21]:
print(Toronto_venues.shape)
Toronto_venues.head()

(2139, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

In [22]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            5   
Alderwood, Long Branch                                               7   
Bathurst Manor, Wilson Heights, Downsview North                     21   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   22   
...                                                                ...   
Willowdale, Willowdale West                                          5   
Woburn                                                               4   
Woodbine Heights                                                     8   
York Mills West                                                      2   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             5   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Willowdale West                                           5   
Woburn                                                                4   
Woodbine Heights                                                      8   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            5               5   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Willowdale West                          5               5   
Woburn                                               4               4   
Woodbine Heights                                     8               8   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      5   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Willowdale West                                    5   
Woburn                                                         4   
Woodbine Heights                                               8   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     5  
Alderwood, Long Branch                                        7 

In [23]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 274 uniques categories.


In [24]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Trail  Train Station  \
0                 0                    0  ...      0              0   
1                 0                    0  ...      0              0   
2                 0                    0  ...      0              0   
3                 0                    0  ...      0              0   
4                 0                    0  ...      0              0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                   0                              0                 0   
1                   0                              0                 0   
2                   0                              0                 0   
3                   0                              0                 0   
4                   0                              0                 0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 274 columns]

In [25]:
Toronto_onehot.shape

(2139, 274)

In [26]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()

In [27]:
Toronto_grouped.shape

(96, 274)

In [28]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [30]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink        Breakfast Spot   
1                   Pharmacy                   Gym        Sandwich Place   
2                       Bank    Frozen Yogurt Shop         Shopping Mall   
3        Japanese Restaurant    Chinese Restaurant                  Bank   
4         Italian Restaurant           Coffee Shop         Grocery Store   

  5th Most Common Venue    6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store      Dumpling Restaurant   Distribution Center   
1           Coffee Shop                      Pub   Distribution Center   
2           Bridal Shop           Sandwich Place                 Diner   
3         Women's Store      Distribution Center               Dog Run   
4       Thai Restaurant  Comfort Food Restaurant             Juice Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run      Doner Restaurant             Donut Shop  
1          Dessert Shop    Dim Sum Restaurant                  Diner  
2            Restaurant         Deli / Bodega            Supermarket  
3      Doner Restaurant            Donut Shop              Drugstore  
4               Butcher            Restaurant                   Café

In [31]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = df

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

Toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
2         M3A        North York                                    Parkwoods   
3         M4A        North York                             Victoria Village   
4         M5A  Downtown Toronto                    Regent Park, Harbourfront   
5         M6A        North York             Lawrence Manor, Lawrence Heights   
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    latitude  longitude  Cluster Labels 1st Most Common Venue  \
2  43.753259 -79.329656             0.0                  Park   
3  43.725882 -79.315572             1.0          Intersection   
4  43.654260 -79.360636             1.0           Coffee Shop   
5  43.718518 -79.464763             1.0        Clothing Store   
6  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2     Food & Drink Shop         Women's Store             Drugstore   
3           Pizza Place           Coffee Shop          Hockey Arena   
4                  Park                   Pub                Bakery   
5         Women's Store           Coffee Shop           Event Space   
6           Yoga Studio      Sushi Restaurant                   Bar   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2          Discount Store   Distribution Center               Dog Run   
3   Portuguese Restaurant            Donut Shop    Dim Sum Restaurant   
4          Breakfast Spot                  Café               Theater   
5  Furniture / Home Store             Gift Shop              Boutique   
6                Beer Bar         Smoothie Shop        Sandwich Place   

  8th Most Common Venue  9th Most Common Venue       10th Most Common Venue  
2      Doner Restaurant             Donut Shop  Eastern European Restaurant  
3                 Diner         Discount Store          Distribution Center  
4                 Hotel         Chocolate Shop                          Spa  
5     Accessories Store  Vietnamese Restaurant            Convenience Store  
6            Restaurant                   Café        Portuguese Restaurant

Some rows have NaN values, so lets drop them ( there are only 3 out of the 103)

In [32]:
Toronto_merged.dropna(inplace=True)

In [39]:
# create map
map_clusters = folium.Map(location=[Tor_lat, Tor_lon], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['latitude'], Toronto_merged['longitude'], Toronto_merged['Neighbourhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Explore Cluster characteristics

In [34]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2          North York             0.0                  Park   
32               York             0.0                  Park   
57          East York             0.0                  Park   
93    Central Toronto             0.0                  Park   
98               York             0.0                  Park   
100        North York             0.0                  Park   
103   Central Toronto             0.0                 Trail   
135       Scarborough             0.0                  Park   
147  Downtown Toronto             0.0                  Park   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2       Food & Drink Shop         Women's Store             Drugstore   
32          Women's Store                  Pool            Donut Shop   
57      Convenience Store          Intersection         Women's Store   
93            Swim School              Bus Line         Women's Store   
98          Women's Store   Dumpling Restaurant        Discount Store   
100     Convenience Store         Women's Store   Dumpling Restaurant   
103                  Park      Sushi Restaurant         Jewelry Store   
135            Playground                Bakery         Women's Store   
147                 Trail            Playground            Donut Shop   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2          Discount Store   Distribution Center               Dog Run   
32     Dim Sum Restaurant                 Diner        Discount Store   
57             Donut Shop                 Diner        Discount Store   
93             Donut Shop        Discount Store   Distribution Center   
98    Distribution Center               Dog Run      Doner Restaurant   
100        Discount Store   Distribution Center               Dog Run   
103            Donut Shop                 Diner        Discount Store   
135             Drugstore        Discount Store   Distribution Center   
147    Dim Sum Restaurant                 Diner        Discount Store   

    8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
2        Doner Restaurant            Donut Shop  Eastern European Restaurant  
32    Distribution Center               Dog Run             Doner Restaurant  
57    Distribution Center               Dog Run             Doner Restaurant  
93                Dog Run      Doner Restaurant                    Drugstore  
98             Donut Shop             Drugstore  Eastern European Restaurant  
100      Doner Restaurant            Donut Shop                    Drugstore  
103   Distribution Center               Dog Run             Doner Restaurant  
135               Dog Run      Doner Restaurant                   Donut Shop  
147   Distribution Center               Dog Run             Doner Restaurant

In [35]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
3          North York             1.0          Intersection   
4    Downtown Toronto             1.0           Coffee Shop   
5          North York             1.0        Clothing Store   
6    Downtown Toronto             1.0           Coffee Shop   
9         Scarborough             1.0  Fast Food Restaurant   
..                ...             ...                   ...   
157  Downtown Toronto             1.0           Coffee Shop   
160         Etobicoke             1.0                 River   
165  Downtown Toronto             1.0           Coffee Shop   
168      East Toronto             1.0            Skate Park   
178         Etobicoke             1.0         Grocery Store   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
3             Pizza Place           Coffee Shop          Hockey Arena   
4                    Park                   Pub                Bakery   
5           Women's Store           Coffee Shop           Event Space   
6             Yoga Studio      Sushi Restaurant                   Bar   
9               Drugstore                 Diner        Discount Store   
..                    ...                   ...                   ...   
157                  Café                 Hotel   Japanese Restaurant   
160                  Pool         Women's Store      Doner Restaurant   
165      Sushi Restaurant   Japanese Restaurant               Gay Bar   
168    Light Rail Station               Butcher      Recording Studio   
178           Social Club                Bakery         Burrito Place   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3     Portuguese Restaurant            Donut Shop    Dim Sum Restaurant   
4            Breakfast Spot                  Café               Theater   
5    Furniture / Home Store             Gift Shop              Boutique   
6                  Beer Bar         Smoothie Shop        Sandwich Place   
9       Distribution Center               Dog Run      Doner Restaurant   
..                      ...                   ...                   ...   
157              Restaurant                   Gym   American Restaurant   
160      Dim Sum Restaurant                 Diner        Discount Store   
165              Restaurant                   Pub           Men's Store   
168           Auto Workshop         Burrito Place                Garden   
178           Tanning Salon          Burger Joint            Kids Store   

        8th Most Common Venue  9th Most Common Venue  10th Most Common Venue  
3                       Diner         Discount Store     Distribution Center  
4                       Hotel         Chocolate Shop                     Spa  
5           Accessories Store  Vietnamese Restaurant       Convenience Store  
6                  Restaurant                   Café   Portuguese Restaurant  
9                  Donut Shop    Dumpling Restaurant          Hardware Store  
..                        ...                    ...                     ...  
157          Asian Restaurant          Deli / Bodega              Steakhouse  
160       Distribution Center                Dog Run              Donut Shop  
165  Mediterranean Restaurant                  Hotel             Yoga Studio  
168             Garden Center         Farmers Market    Fast Food Restaurant  
178            Sandwich Place         Discount Store  Thrift / Vintage Store  

[86 rows x 12 columns]

In [36]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
17  Etobicoke             2.0            Print Shop         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
17                 Diner        Discount Store   Distribution Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
17               Dog Run      Doner Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
17             Drugstore           Dessert Shop

In [37]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
18   Scarborough             3.0  Construction & Landscaping   
77    North York             3.0                        Park   
169    Etobicoke             3.0              Baseball Field   

          2nd Most Common Venue       3rd Most Common Venue  \
18                          Bar               Women's Store   
77                       Bakery  Construction & Landscaping   
169  Construction & Landscaping               Women's Store   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
18    Dumpling Restaurant   Distribution Center               Dog Run   
77          Women's Store   Dumpling Restaurant   Distribution Center   
169   Distribution Center               Dog Run      Doner Restaurant   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
18       Doner Restaurant            Donut Shop             Drugstore   
77                Dog Run      Doner Restaurant            Donut Shop   
169            Donut Shop             Drugstore   Dumpling Restaurant   

          10th Most Common Venue  
18   Eastern European Restaurant  
77                     Drugstore  
169  Eastern European Restaurant

In [38]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
89  North York             4.0        Baseball Field   Dumpling Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
89        Discount Store   Distribution Center               Dog Run   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
89      Doner Restaurant            Donut Shop             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
89         Women's Store   Fast Food Restaurant

It seams that the first 2 clusters concentrate almost all of the values, the other 3 clusters are a lot more unique.